### ISSUE TITLE
Integrate sub-pixel morphological antialiasing (SMAA) in the Vulkan renderer

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#2779]
*Related to https://github.com/godotengine/godot-proposals/issues/3401.*

### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

Right now, we have MSAA and FXAA available in both `3.x` and `master`. These algorithms serve their intended purposes, but each of them has their own limitations:

- MSAA looks good, but it is very expensive on `master` due to the added complexity of shaders. This is especially noticeable when using GIProbe or SDFGI. It also doesn't smooth out aliasing that originates from fragment shaders such as specular aliasing. (Alpha-tested surfaces are now well-handled thanks to [alpha antialiasing and alpha-to-coverage support](https://github.com/godotengine/godot/pull/40364).)
- FXAA is cheap and looks decent on 1440p and higher. Unfortunately, the added blurriness makes it hard to use on 1080p and below. [Post-processing sharpening algorithms](https://github.com/godotengine/godot-proposals/issues/1519) make it possible to recover some of that lost sharpness, but they also add more aliasing on their own, so it's a tradeoff.

It's possible to use both MSAA and FXAA at the same time, but this has a significant cost and it won't improve the image's sharpness after it has been reduced by FXAA.

It's great that MSAA is still supported in `master` (thanks to the clustered forward renderer), but we have to admit that it's difficult to use in production with today's shader complexity. Therefore, this makes MSAA mostly suited to games that use few visual effects, which generally implies a stylized art direction.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

[Sub-pixel morphological antialiasing (SMAA)](https://www.iryoku.com/smaa/) has been around since 2011 and is now a proven post-processing-based antialiasing algorithm. It's still used in a lot of AAA games.

Performance-wise, SMAA is more expensive than FXAA, but it does a good job at antialiasing despite introducing less blurriness. Either way, SMAA is likely to be less expensive than even 2× MSAA on the Vulkan renderer (despite providing better edge smoothing overall).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

One downside of SMAA is that it's notoriously difficult to implement. Not only it's a multi-pass algorithm (unlike FXAA which is single-pass), there are also many quality settings to tinker with.

We can probably go with the recommended SMAA settings (luma-based edge detection), and only see if there are real benefits to exposing quality knobs in the project settings (such as changing the edge detection algorithm).

SMAA also offers an optional temporal component, but that comes with added complexity and motion trails commonly associated with temporal antialiasing algorithms. Therefore, I suggest we leave out the temporal antialiasing part of SMAA and focus on spatial-only SMAA (also called SMAA 1×). SMAA won't look as smooth without this temporal component, but SMAA still looks better than FXAA without it.

I've looked around a bit and couldn't figure out how to add a multi-pass post processing shader in Godot's GLSL core, so feel free to give it a try.

**Question: Can this be implemented in `3.x`?**
~~I haven't seen any implementation of SMAA on OpenGL ES 3.0 (and even less OpenGL ES 2.0), so probably not.~~
**Edit:** It's possible to implement SMAA in WebGL 2.0 (and therefore OpenGL ES 3.0): https://github.com/dmnsgn/glsl-smaa

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Since [real-time multi-pass shaders aren't possible in "userland" Godot yet](https://github.com/godotengine/godot-proposals/issues/496), no.

### Is there a reason why this should be core and not an add-on in the asset library?

See above.

**Comment by clayjohn** [comment]
For reference, the original SMAA code can be found [here](https://github.com/iryoku/smaa) and it is MIT licenced.

**Comment by AlfonsoGrimaldo** [comment]
I found this: [https://github.com/dmnsgn/glsl-smaa](url), i hope it helps.

**Comment by Calinou** [comment]
This porting work may also be useful as a reference: https://github.com/libretro/slang-shaders/pull/189

**Comment by jntesteves** [comment]
> This porting work may also be useful as a reference: [libretro/slang-shaders#189](https://github.com/libretro/slang-shaders/pull/189)

Hey, if you want to take any reference from my work there, please check the aftermath: [libretro/slang-shaders#192](https://github.com/libretro/slang-shaders/pull/192)

I can't agree with the docs on luma by default. I tested it extensively and color based edge detection always gave me better results, with less aliasing but not blurrier. In the end, I exposed every config knob, so I could more easily tweak settings at runtime. Tweaking it for your content is worthwhile, this AA is powerful when properly setup.

`SMAA_PRESET_ULTRA` + `SMAAColorEdgeDetectionPS` seemed like a good default to me. I increased AA strength a nod (`SMAA_LOCAL_CONTRAST_ADAPTATION_FACTOR 2.5`) where I knew there would be sharpening in a following pass, like SMAA+FSR.

My code is in the public domain, no restrictions apply. Feel free to copy anything.

**Comment by Calinou** [comment]
I've started work on integrating SMAA: https://github.com/Calinou/godot/tree/add-smaa-antialiasing

It's far from functional yet, but feel free to take a look nonetheless.

**Comment by Lasuch69** [comment]
Is there any update on this form of anti-aliasing?

**Comment by Calinou** [comment]
> Is there any update on this?

I haven't managed to get it working, as I have very little time to work on this.

**Comment by jntesteves** [comment]
I don't have any stake in Godot at the moment, so my opinion might not be too relevant, but I think working on the multi-pass shaders proposal mentioned in the description is a more worthwhile use of time. With that, anyone could easily do SMAA or whatever other post-process effects they want, not limited by what the engine offers.

**Comment by Jamsers** [comment]
CMAA2 may be a great alternative to consider, it's been used to great effect in the recently released Counter Strike 2 and seems to work well as an MSAAesque cheap post process fallback.

The original repo published by Intel: https://github.com/GameTechDev/CMAA2
A ReShade implementation: https://github.com/LordOfLunacy/Insane-Shaders/blob/master/Shaders/CMAA_2.fx

**Comment by jams3223** [comment]
How is this coming along, i was going to open a proposal until i saw this one, SMAA is the perfect  middle ground in term of performanve, quality and artifacts.

Q: Describe the project you are working on
A: 3D games that make use of realistic art styles and complex shapes

Q: Describe the problem or limitation you are having in your project
A: TAA tends to create ghosting artifacts while in motion, and while TAA is better at removing aliasing than SMAA, it still does a good job without having these issues; it's a good middle ground between TAA and FXAA. TAA is more resource-intensive and prone to artifacts, while FXAA is less effective.

Q: Describe the feature / enhancement and how it helps to overcome the problem or limitation
A: Add it to the list of already-available anti-aliasing algorithms.

Q: If this enhancement will not be used often, can it be worked around with a few lines of script?
A: No, it can't be used with a few lines of code; it needs to be implemented into the core engine.

Q: Is there a reason why this should be core and not an add-on in the asset library?
A: It sits right along with the other anti-aliasing methods that Godot ships with, so it helps extend the user's preferred choice and ensure a balanced level of quality.

**Comment by Calinou** [comment]
> How is this coming along? I was going to open a proposal until I saw this one. SMAA is the perfect middle ground in terms of performance, quality, and artifacts.

SMAA is generally considered to be a stale approach by 2023 standards, so I wouldn't hold my hopes for it being added as a core feature (in the age of ever-improving TAA solutions). However, the rendering engine will eventually be made more flexible so that SMAA can be implemented by an add-on.

**Comment by Jamsers** [comment]
>Modern games are often full of geometric and specular detail that SMAA can't do anything about

That's fine - the users wanting to use SMAA aren't expecting it to match or even come close to the antialiasing capability of TAA - the point is to get the best possible antialiasing possible today that isn't TAA (due to ghosting and artifacting that will never be solved no matter how improved TAA gets) or MSAA (due to prohibitive performance cost)

>SMAA is generally considered to be a stale approach by 2023 standards

Irrelevant. Progress and development on non-temporal post process antialiasing methods has practically halted, so unless the Godot team is planning to create a bespoke new technique, SMAA is still the best there is. (again, if you don't want TAA or MSAA)

If anything, FXAA is the stale, outdated, irrelevant technique - why is *that* still in core?

**Comment by jams3223** [comment]
I have tested the anti-aliasing algorithm with every kind of game art style, and I can say that it all depends on the art style or visual style. Godot will benefit by giving people choices in how they want to customize their game.

FXAA:
+ Better with cel-shaded and anime art styles.
- Lower Quality
+ Better Performance

MSAA:
+ Better with realistic art styles.
+ Higher Quality
- Lower Performance

TAA:
+ Better with realistic art styles.
+ Higher Quality
+ Balanced Performance
- Visual Artifact

SMAA:
+ Better with realistic, anime and voxel art styles.
+ Balanced Quality
+ Balanced Performance

**Comment by mrjustaguy** [comment]
One thing I see forgotten in here is the fact that SMAA can optionally be temporal itself, also known as SMAAT2x

**Comment by mrjustaguy** [comment]
SMAAT2x still beats our TAA implementation in quality and likely even performance (though that is largely because of just how totally borked it is) and FSR2 is really expensive at native res, though I don't know the diff between SMAAT2x and FSR2

**Comment by Jamsers** [comment]
Absolutely, but as you say, that's a testament to how terrible Godot's TAA implementation is, rather than SMAA T2x being the way to go. And the current FSR 2 implementation has a disproportionately high performance cost - it's [damn near 12% of GPU time](https://github.com/Jamsers/Bistro-Demo-Tweaked/issues/29) on a relatively heavy and complex scene - so some optimization work is definitely due there.

**Comment by mrjustaguy** [comment]
FSR2 performance matches what AMD states for it's Performance Costs, and it's really designed for Upscaling, where you're rendering at a lower internal resolution (reducing the FSR2 cost) and Upscaling, improving the performance due to significantly less shading per frame

**Comment by Xynonners** [comment]
Hey all,

just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.

here is my LXAA port:
https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93
and here is HHAA (a weird mishmash bespoke algorithm I made for 3D):
https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623

someone smarter than me could probably do a CMAA2 port if they wanted to.

pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).

**Comment by jams3223** [comment]
> Hey all,
> 
> just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> 
> here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> 
> someone smarter than me could probably do a CMAA2 port if they wanted to.
> 
> pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).

I've never heard of HHAA, did you make it yourself?

**Comment by Xynonners** [comment]
> > Hey all,
> > just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> > here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> > someone smarter than me could probably do a CMAA2 port if they wanted to.
> > pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).
> 
> I've never heard of HHAA, did you make it yourself?

Yup, the idea was to be more aggressive than LXAA (which tends to be extremely conservative for the sake of anti-blur), but less blurry than the blur-fest that is FXAA. HHAA is a weird hybrid/combined antialiasing shader, but imo it works pretty well (though I do believe it could still be more aggressive - though I don't think I'm smart enough to make something better than this). 

Because of it's completely custom nature though, it may have a sizeable performance impact, though I roughly tested it and it seems to be fine.

**Comment by Jamsers** [comment]
It produces pretty SMAAesque results to my eyes when I tested it on Bistro-Demo-Tweaked. Any chance you could port CeeJayDK's SMAA? 👉👈🥺

I tried it myself but I have zero shader knowledge and only got as far as porting the UI variables 😅

**Comment by Beanibirb** [comment]
> Hey all,
> 
> just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> 
> here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> 
> someone smarter than me could probably do a CMAA2 port if they wanted to.
> 
> pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).
Inexperienced with setting up shaders for Godot here, how would one implement either of these in a 3D template?

**Comment by mrjustaguy** [comment]
See https://docs.godotengine.org/en/stable/tutorials/shaders/custom_postprocessing.html

**Comment by Beanibirb** [comment]
> Hey all,
> 
> just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> 
> here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> 
> someone smarter than me could probably do a CMAA2 port if they wanted to.
> 
> pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).

Tested the two out, HHAA is good but you were right about it being very aggressive as it does not seem to like particle effects whatsoever.

![Screenshot (133)](https://github.com/godotengine/godot-proposals/assets/26067418/2e2e628d-423c-4b3a-a553-6b8567d92a3d)

This is using the default values for the hhaa-d shader.

**Comment by Xynonners** [comment]
> > Hey all,
> > just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> > here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> > someone smarter than me could probably do a CMAA2 port if they wanted to.
> > pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).
> 
> Tested the two out, HHAA is good but you were right about it being very aggressive as it does not seem to like particle effects whatsoever.
> 
> ![Screenshot (133)](https://private-user-images.githubusercontent.com/26067418/313314404-2e2e628d-423c-4b3a-a553-6b8567d92a3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1ODczNDIsIm5iZiI6MTcxMDU4NzA0MiwicGF0aCI6Ii8yNjA2NzQxOC8zMTMzMTQ0MDQtMmUyZTYyOGQtNDIzYy00YjNhLWE1NTMtNmI4NTY3ZDkyYTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDExMDQwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxZWY1ZmZjNTQ2YjEyOTU3YjRjZGJkZDdmODVkOGUwZGYwM2FmZjI3NTBhZWM2ZDdiYjc4Nzk1N2VmNmRlNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wfU5A36jdRs9jumOHC9TM895MroXNeliD5i64aPVYq0)
> 
> This is using the default values for the hhaa-d shader.

yeah, after making HHAA-D, I did a whole bunch more testing and ironed out the bugs. The edge detection in HHAA originally had a ton of false positives that were being dropped later on by the blending technique.

I made an improved version called AHAA (actually being used in a shipgame now):
https://gist.github.com/Xynonners/0edfc4ce5cab61f545041752e53b3f5b

**Comment by Beanibirb** [comment]
> > > Hey all,
> > > just dropping in to add that it is actually possible to implement / port reshade AA to gdshader.
> > > here is my LXAA port: https://gist.github.com/Xynonners/6e50a27d562829469281be45b039ec93 and here is HHAA (a weird mishmash bespoke algorithm I made for 3D): https://gist.github.com/Xynonners/ddf30a5215a59db5690f6420621f0623
> > > someone smarter than me could probably do a CMAA2 port if they wanted to.
> > > pretty tired of waiting for AA from godot, considering that FXAA is stale, TAA is broken (motion smear).
> > 
> > 
> > Tested the two out, HHAA is good but you were right about it being very aggressive as it does not seem to like particle effects whatsoever.
> > ![Screenshot (133)](https://private-user-images.githubusercontent.com/26067418/313314404-2e2e628d-423c-4b3a-a553-6b8567d92a3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1ODczNDIsIm5iZiI6MTcxMDU4NzA0MiwicGF0aCI6Ii8yNjA2NzQxOC8zMTMzMTQ0MDQtMmUyZTYyOGQtNDIzYy00YjNhLWE1NTMtNmI4NTY3ZDkyYTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDExMDQwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxZWY1ZmZjNTQ2YjEyOTU3YjRjZGJkZDdmODVkOGUwZGYwM2FmZjI3NTBhZWM2ZDdiYjc4Nzk1N2VmNmRlNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wfU5A36jdRs9jumOHC9TM895MroXNeliD5i64aPVYq0)
> > This is using the default values for the hhaa-d shader.
> 
> yeah, after making HHAA-D, I did a whole bunch more testing and ironed out the bugs. The edge detection in HHAA originally had a ton of false positives that were being dropped later on by the blending technique.
> 
> I made an improved version called AHAA (actually being used in a shipgame now), maybe that would be better? https://gist.github.com/Xynonners/0edfc4ce5cab61f545041752e53b3f5b
> 
> also, another thing is that it is **VERY** important to set the render_priority of the ShaderMaterial to -128 (also causes the noise issue).

The render_priority was the only thing that was messing me up, thank you so very much for this

**Comment by Jamsers** [comment]
The recent fantastic progress on implementing motion blur as a Compositor effect has got me thinking - how hard would it be to implement SMAA as a compositor effect? ᵃⁿᵈ ʷᵒᵘˡᵈ ᵃⁿʸᵒⁿᵉ ᵇᵉ ʷⁱˡˡⁱⁿᵍ ᵗᵒ ⁱᵐᵖˡᵉᵐᵉⁿᵗ ⁱᵗ ᵖˡᵉᵃˢᵉ 😁

**Comment by Calinou** [comment]
> how hard would it be to implement SMAA as a compositor effect?

I think it's within the realm of possibility, as multipass effects can be implemented this way.

It's probably easier than motion blur in fact, since SMAA only depends on the color buffer (at least in its higher quality modes).

**Comment by RGDTAB** [comment]
Just made an SMAA compositor effect [here](https://github.com/RGDTAB/Godot-SMAA). It's only SMAA 1x, but I plan on working on T2x soon.

**Comment by Jamsers** [comment]
Raymond there's nothing else to say, you're based, you're a chad, you are him, and all the other Gen Z superlatives I can muster 💖💖💖

I've tested the compositor effect and it's perfect. Looks exactly as you'd expect, works at differing render scales, costs pretty much nothing in terms of performance.

For anyone else looking to use RGDTAB's compositor effect, I heavilly recommend setting the edge detection method to color and the quality to at least high. (medium and below disables diagonal and corner detection, which has a disproportionate cost to quality compared to the meager performance benefits)

**Comment by mrjustaguy** [comment]
Color isn't the best, it'll introduce Moire artifacts on some objects with High frequency textures:

SMAA Color Ultra (same happens on all other settings)
![image](https://github.com/user-attachments/assets/f6be4905-fd5f-4025-b2e9-41522e4a9ed7)

SMAA Luma Ultra (same on all other settings)
![image](https://github.com/user-attachments/assets/f179efd6-6b08-4df6-ae09-b7f39abc80da)

As the person who made a PR (https://github.com/godotengine/godot/pull/89582) to update Godot's FXAA to Nvidia's FXAA version 3.11 I'm glad you've made that essentially obsolete for Godot IMHO, and I'm looking forward to seeing what other modes of SMAA will be like in Godot

**Comment by Jamsers** [comment]
Moire on patterns is an expected (and acceptable) side effect of full morphological AA - the alternative would be not touching certain parts of the image, which would allow aliasing.

IIRC both Jorge Jimenez (the creator of SMAA) and Christian Jensen (implemented SMAA for SweetFX) have confirmed that Luma was always meant to be a less accurate but faster mode, and is pretty much just a legacy leftover from the old days. Using Luma in actual game scenarios, you'll miss out on anti-aliasing texture detail, foliage, and even edges within shadowed areas.

In SweetFX, which would pretty much be the reference implementation nowadays since games stopped using SMAA, the default is Color and Ultra - but with threshold set to 0.1 rather than 0.05.

**Comment by Jamsers** [comment]
Not the shadows: objects and surfaces *in* shadow. Basically scenarios where there's low luminance contrast and only high color contrast. This was more relevant back in the old days where lots of open world games only had IBL as ambient light as opposed to the fancier bounce lighting we tend to have nowadays, but is fundamentally relevant to the difference of the two approaches.

Luma:

![luma_medium](https://github.com/user-attachments/assets/6ad971d1-24f1-4fea-a682-45a491620900)

Color:

![color_medium](https://github.com/user-attachments/assets/494f0fb5-c5a1-4f5c-b9dd-cfb30bc57bc5)

Open the images in fullscreen, note the boundary between tree bark and tree leaves, and the grass.

**Comment by mrjustaguy** [comment]
as far as SMAAT2x is concerned, jittering the camera could be a problem, as Camera3D doesn't allow you to modify the Projection, although it Does allow you to get it using `get_camera_projection` which even has a `jitter_offset` that you could use on the projection, there's just seemingly no way to feed it back to the camera, and the other option of just changing raw position of the camera could pose issues if the player is modifying the camera position themselves..

**Comment by RGDTAB** [comment]
Yeah, you're right. It looks like there's an open PR to allow custom Camera3D projection matrices, so once that gets merged I'll start work on it.
In the meantime I'll try and tackle SMAAS2x. It should be possible in theory, but we'll see.

**Comment by Beanibirb** [comment]
> Just made an SMAA compositor effect [here](https://github.com/RGDTAB/Godot-SMAA). It's only SMAA 1x, but I plan on working on T2x soon.

This is actually amazing dude THANK YOU

**Comment by ShalokShalom** [comment]
I come here based on one single Youtube comment - so beware. 😄 

It said, that SMAA is the only viable option for VR. After reading that threat, I somehow question it, although I have to say, TAA seems not a good option for this use case for sure.

What do we use - and recommend for VR?
Is SMAA a good option for that?

Alternatives, that we haven't implemented yet?

**Comment by Jamsers** [comment]
TAA is 100% bad for VR, you are going to make your players sick, guaranteed, doesn't matter if it's really sharp or fancy like DLSS.

SMAA will be better for VR, but still fundementally flawed, since you'll still get subpixel "dancing" and general inconsistency between the two renders. This'll be true for any post process antialiasing solution.

The only viable options for VR are MSAA... and SSAA I guess, but I don't think it's practical to require a supercomputer from your players 😅

**Comment by mrjustaguy** [comment]
FXAA and SMAA are the only viable options for VR. 
TAA of any kind is a horrible idea both because of how demanding it is relative to FXAA and SMAA (also why MSAA and SSAA Aren't good too, too expensive) and TAA introduces quite a few artifacts that will give players headaches. 
It's another reason why Post Processing in general (like ssr and ssao, sss, sscs and what not) is very bad for VR, as it gives different results with different artifacts from eye to eye, which is bad.

Now FXAA and SMAA also give slightly different results from eye to eye, but they're on the level of "so does no antialiasing, you just have to live with it" type of deal, and is somewhat mitigated by the ultra high resolutions made possible by eye tracking and VRS

**Comment by Jamsers** [comment]
On mobile hardware based VR platforms like the Quest, MSAA is not as deleteriously expensive as it is on PC/console, and is more practical. (Although still expensive)

On PC VR, you'll either follow in the footsteps of Half Life Alyx and make MSAA work at all costs and have players enjoy your game, or you'll go the Bethesda route and deploy with TAA or FXAA and turn players off VR forever.

There's no easy path - the harsh reality with VR right now is if you can't afford MSAA with your project, then you can't afford your project, period.

**Comment by ShalokShalom** [comment]
It seems like Unreal, Unity, and a bunch of others add a specially adjusted TAA for VR.

But I dont have any experience with it. To me, this reads like a custom AA for VR is suitable. 
Considering that even Valve has choosen to go with MSAA instead, I believe doing such a custom AA from scratch is out of scope for the next years. 

Khornos is the only entity, I see capable doing that.

**Comment by Calinou** [comment]
> It said, that SMAA is the only viable option for VR. After reading that threat, I somehow question it, although I have to say, TAA seems not a good option for this use case for sure.

The person writing you probably meant MSAA instead of SMAA. These are easily confused for each other but are very different. MSAA is multisample antialiasing (**not** post-process AA) while SMAA is subpixel morphological antialiasing (post-process AA).

**Comment by ShalokShalom** [comment]
I see. 

Adjusting the AA for VR, like to accomodate for edge detection, specifically tuned for VR lens distortion and implementing per-eye processing, is something done in any of the AA in Godot?

**Comment by Calinou** [comment]
> Adjusting the AA for VR, like to accomodate for edge detection, specifically tuned for VR lens distortion and implementing per-eye processing, is something done in any of the AA in Godot?

Not that I know of, but I wouldn't use FXAA in VR due to its blurriness.

**Comment by ShalokShalom** [comment]
But that should apply to TAA as well, right?

**Comment by Calinou** [comment]
> But that should apply to TAA as well, right?

Yes, TAA has similar problems with VR (on top of ghosting issues).

**Comment by ShalokShalom** [comment]
> most of the aliasing that's distracting in VR is of a subpixel nature (and post-process AA can't do anything about it).

Just to be sure, pancake optics can provide clearer edge definition, so they are less susepticle to that?

And considering that higher resolution displays spread the effect across more pixels, I assume that makes individual subpixel patterns less noticeable?

It would also be more noticable on OLEDs, considering their uneven subpixel distribution?

**Comment by Calinou** [comment]
> Just to be sure, pancake optics can provide clearer edge definition, so they are less susepticle to that?

While pancakes look sharper to the human eye, the underlying issues with post-processing based AA blur being noticeable are still there.

Besides, it's not like modern VR displays are *too* sharp right now. Even my Quest 3 with 1.5x supersampling on each axis still looks a bit blurry, and that's *without* any lens inserts (I wear glasses with slight correction normally). To get a truly sharp image in VR for a majority of users (where pixels are impossible to distinguish from each other), you'd need 8K resolution.
